R/last_fit.R

```diff
@@ -65,42 +69,66 @@ last_fit <- function(object, ...) {
 last_fit.default <- function(object, ...) {
   empty_ellipses(...)
   msg <- paste0(
-    "The first argument to [last_fit()] should be either a ",
-    "formula, recipe, or workflow."
+    "The first argument to [last_fit()] should be either ",
+    "a model or workflow."
   )
   rlang::abort(msg)
 }
 
 #' @rdname last_fit
 #' @export
 last_fit.recipe <- function(object, model, split, ..., metrics = NULL) {
+  lifecycle::deprecate_soft("0.1.0",
+                            what = "last_fit.recipe()",
+                            details = deprecate_msg(match.call(), "last_fit"))
   empty_ellipses(...)
-  if (is_missing(model) || !inherits(model, "model_spec")) {
-    rlang::abort("`model` should be a parsnip model specification object.")
-  }
 
-  workflow <- workflow()
-  workflow <- add_recipe(workflow, object)
-  workflow <- add_model(workflow, model)
+  last_fit(model, preprocessor = object, split = split, metrics = metrics)
 
-  last_fit_workflow(workflow, split, metrics)
 }
 
 #' @rdname last_fit
 #' @export
 last_fit.formula <- function(formula, model, split, ..., metrics = NULL) {
+  lifecycle::deprecate_soft("0.1.0",
+                            what = "last_fit.formula()",
+                            details = deprecate_msg(match.call(), "last_fit"))
   empty_ellipses(...)
-  if (is_missing(model) || !inherits(model, "model_spec")) {
-    rlang::abort("`model` should be a parsnip model specification object.")
+
+  last_fit(model, preprocessor = formula, split = split, metrics = metrics)
+}
+
+#' @export
+#' @rdname last_fit
+last_fit.model_spec <- function(object, preprocessor, split, ..., metrics = NULL) {
+
+  if (is_missing(preprocessor) ||
+      !(inherits(preprocessor, "recipe") ||
+        inherits(preprocessor, "formula"))) {
+    rlang::abort(paste("To tune a model spec, you must preprocess",
+                       "with a formula or recipe"))
   }
 
-  workflow <- workflow()
-  workflow <- add_formula(workflow, formula)
-  workflow <- add_model(workflow, model)
+  empty_ellipses(...)
+
+  wflow <-
+    workflow() %>%
+    add_model(object)
```

Comment:
Stylistically, I'm _personally_ team anti-pipe when I write package code unless it really helps. It has a very minor performance hit, can sometimes make the tracebacks a little harder to read if something does error, and for one-liners I often find the non-piped equivalent a little easier to read. I think Max feels differently though so its fine to keep as is

Reply:
I'm with you on this, especially for one-liners. Addressed in db74821d2a0c096ca8f1075349acf74be8bb7dde